[docker_network] add ipv6 support

smueller18 opened the pull request:
##### SUMMARY
This PR enables support for IPv6 networks and allows defining IPv6 IPAM config blocks. The configuration options are based on the existing docker-compose documentation https://docs.docker.com/compose/compose-file/compose-file-v2/#ipam .

So far, the parameter for IPAM config was named `ipam_options` but in recent docker versions, this is called `ipam_config` (see link above). Here is the example from the docker-compose docs:
```
ipam:
  driver: default
  config:
    - subnet: 172.28.0.0/16
      ip_range: 172.28.5.0/24
      gateway: 172.28.5.254
      aux_addresses:
        host1: 172.28.1.5
        host2: 172.28.1.6
        host3: 172.28.1.7
  options:
    foo: bar
    baz: "0"
```
So I changed the parameter name `ipam_options` to `ipam_config`. I also created a branch (https://github.com/smueller18/ansible/tree/docker_network_ipam_options) that handles the new parameter `ipam_options` well, but this change requires docker >= 2.0.0, so I cannot create a pull request for that change until support for docker-py and python 2.6 is dropped.

Fixes #34855

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
docker_network

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```paste below
ansible 2.8.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/sm/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/sm/git/github/smueller18/ansible/venv/lib/python3.7/site-packages/ansible-2.8.0.dev0-py3.7.egg/ansible
  executable location = /home/sm/git/github/smueller18/ansible/venv/bin/ansible
  python version = 3.7.0 (default, Aug 30 2018, 22:41:14) [GCC 7.3.0]
```

smueller18 committed:
[docker_network] add ipv6 support

DBendit commented:
Since you're adding a new parameter and renaming an existing one, you'll need to add a [changelog fragment](https://groups.google.com/forum/#!msg/ansible-devel/uYuIADYa5vs/Cby5dnDRAAAJ;context-place=forum/ansible-devel).

needs_revision

DBendit commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -62,13 +62,22 @@
     description:
       - Specify an IPAM driver.
 
-  ipam_options:
```
This is a breaking change to existing functionality. Rather than outright replacing the option here, why not create `ipam_config` as a new option and include logic to use it preferentially if it's set? That way, upgrading users won't have their playbooks break immediately.

Additionally, the `ipam_options` option should be marked as deprecated when doing so.

felixfontein replied:
Simply removing an existing option is not acceptable. Adding a new option and deprecating the old one is the best way to do this correctly. (Also, you should include some logic which forbids to use both.)

smueller18 replied:
I see your point. This is a bit sad, because the parameters then differ from the official docker specification. So I will add a deprecation note for `ipam_options` and implement the "real" ipam options as parameter `ipam_driver_options`.

felixfontein replied:
Why do the parameters differ?

You can still name the new parameter `ipam_options`, as long as you keep the old one (marked as deprecated, to be removed in Ansible 2.12) and make sure that specifying both doesn't lead to problems (i.e. simply fail if both are specified). Internally, you can convert the old one to the new one if the old one is specified.

smueller18 replied:
In Docker 1.10.0 (https://docs.docker.com/release-notes/docker-engine/#1100-2016-02-04, https://github.com/moby/moby/pull/17316) the new parameter `--ipam-opt` was introduced. But because the docker_network module already defined the parameter `ipam_options` (which actually is called ipam config in the docker specification) I cannot use `ipam_options` for the newly introduced parameter.

felixfontein replied:
Ok, I was mainly confused because a third (not yet mentioned) name showed up. So the idea is to:
 1. deprecate `ipam_options`;
 2. add `ipam_config`, which is auto-created from `ipam_options` if that one is specified;
 3. add `ipam_driver_options` (for the IPAM driver-specific options).
That sounds OK for me.

Otherwise, we could maybe also proceed as follows: deprecate `ipam_driver` and `ipam_options`, and add `ipam` parameter, which accepts something like [here](https://docs.docker.com/compose/compose-file/compose-file-v2/#ipam). This requires a lot more error-checking on the module's side, though, since Ansible won't do much (except guaranteeing that `ipam` is a `dict`).

smueller18 replied:
I tried to implement the parameter `ipam_driver_options` (requires docker-py >= 2.0.0), but it requires a lot of changes to the docker_network module to guarantee backward compatibility. So I decided not to implement this feature.

DBendit commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -308,6 +358,7 @@ def create_network(self):
                                                   driver=self.parameters.driver,
                                                   options=self.parameters.driver_options,
                                                   ipam=ipam_config,
+                                                  enable_ipv6=self.parameters.enable_ipv6,
```
The `enable_ipv6` option was added in 1.10.0 of docker-py (see [changelog](https://docker-py.readthedocs.io/en/stable/change-log.html)). You'll need to either conditionally enable this parameter (see `docker-container` for examples of how to do this), or update the minimum version of docker-py required (in both the documentation above and the `AnsibleDockerClient` initialization below).

smueller18 replied:
Is it acceptable to set requirements for the whole docker_network to python >= 2.7 and docker-py >= 2.0.0 ? For me, this is the preferred method.

felixfontein replied:
No, I don't think that's a good idea.

felixfontein commented:
Also (besides the comments by @DBendit), note that there's already a PR for this: #34509 by @huiyiqun. It hasn't been updated for some time though, and I think the approach in this PR is better (having a new option which allows to specify multiple IPAM configs).

smueller18 committed:
docker_network: review ipam_options

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -395,14 +468,16 @@ def main():
         appends=dict(type='bool', default=False, aliases=['incremental']),
         ipam_driver=dict(type='str', default=None),
         ipam_options=dict(type='dict', default={}),
```
Please add `removed_in_version='1.12'` to `ipam_options`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -62,13 +62,35 @@
   ipam_options:
     description:
       - Dictionary of IPAM options.
+    deprecated:
```
Please remove this, and add an item with the deprecated explanation to `description`. Also, please use correct casing for `IPAM`.

smueller18 committed:
docker_network: fix requirements

smueller18 committed:
docker_network: fix deprecation notice

smueller18 committed:
docker_network: add minimum docker version change

smueller18 committed:
docker_network: remove trailing whitespace

felixfontein commented:
@DBendit I'm not sure whether I'll have a chance to take a good look at it today; otherwise I'll check over the weekend.

felixfontein commented on changelogs/fragments/47492-docker_network-add-ipv6-support.yaml:
```diff
@@ -0,0 +1,6 @@
+minor_changes:
+  - "docker_network - Add support for ipv6 networks."
+  - "docker_network - Minimum docker-py version increased from ``1.9.0`` to ``1.10.0``."
```
If we really want to bump the docker-py version again (we've already done it once for 2.8 from 1.8.0 to 1.9.0 in #35370), we should not have two changelog entries (1.8 -> 1.9 and 1.9 -> 1.10), but merge them. The existing one is in https://github.com/ansible/ansible/blob/devel/changelogs/fragments/35370-add_support_for_docker_network_internal_flag.yaml

smueller18 replied:
How should I merge them? Delete the second entry of #35370 and set my entry to 
```"docker_network - Minimum docker-py version increased from ``1.8.0`` to ``1.10.0``."``` ?

felixfontein replied:
Yes, that's probably the best way. Or create a new changelog fragment `docker_network-requirements.yaml` which contains that one, and remove it from both yours and the other changelog fragment. In that case, there's one unique place where it can be bumped in the future.

smueller18 replied:
merged it into the newly proposed file name

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -62,13 +62,32 @@
+  ipam_config:
+    version_added: 2.8
+    description:
+      - List of IPAM config blocks. Consult docker docs for valid options and values.
```
You should point to the docker documentation here, so that users know which options and values they can use.

smueller18 replied:
Done

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -62,13 +62,32 @@
+    description:
+      - List of IPAM config blocks. Consult docker docs for valid options and values.
+    type: list
+    default: []
```
Please don't add a default of `[]`. This makes it impossible to distinguish between "the user didn't specify options" and "the user explicitly specified no options" (this distinction is important for idempotency checks, even though not right now, but it might block future additions similar to the `comparisons` option for `docker_container`).

smueller18 replied:
Most of the other options (e.g. `connected`) also have an empty list as default so I did the same. Changing to `None`.

felixfontein replied:
`connected` is a bit different, since it is not an option passed to docker, but something processed completely by this module. I would have not used defaults for some of the others, but that's now too late to change (to keep backwards compatibility).

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -141,19 +160,38 @@
 - name: Delete a network, disconnecting all containers
   docker_network:
-    name: network_one
+    name: network_four
```
I don't think you defined `network_four` above, so why not keep `network_one` here?

smueller18 replied:
Yes, network_four was never defined, reverting to network_one

felixfontein commented on changelogs/fragments/47492-docker_network-add-ipv6-support.yaml:
```diff
@@ -0,0 +1,6 @@
+minor_changes:
+  - "docker_network - Add support for ipv6 networks."
```
The correct spelling is `IPv6`, not `ipv6`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -141,19 +160,38 @@
+        gateway: 172.3.27.1
+        iprange: 192.168.1.0/24
+
+- name: Create a network with ipv6 IPAM config
```
Please also use IPv4 and IPv6 instead of their lowercase variants in the comment/names (here and below).

smueller18 replied:
ok

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -219,6 +261,11 @@ def __init__(self, client):
             self.parameters.connected = container_names_in_network(self.existing_network)
 
+        if self.parameters.ipam_options and self.parameters.ipam_config:
+            self.client.fail("Only one of `ipam_options` and `ipam_config` can be defined.")
```
Don't do this check manually, but let Ansible do it. (Search for `mutually_exclusive` in docker_service.py)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -253,29 +300,50 @@ def has_different_config(self, net):
+                            net_config = dict()
+                        # ipv6 already exists
+                        else:
+                            net_config = net['IPAM']['Config'][1]
```
Has `net['IPAM']['Config']` always one or two entries, with IPv4 always being the first and IPv6 always being the second?

smueller18 replied:
Yes, IPv4 is always appended first, see https://github.com/moby/moby/blob/cf72051c3746fa60d7e7995c0342eaca91f07bfd/daemon/network.go#L675

smueller18 replied:
But there is a feature request for IPv6 only network (https://github.com/moby/moby/issues/37798). If this is implemented, the current logic would not work. Adjusting this code to work independently from index.

felixfontein replied:
It might be possible to convert the existing values to a syntax simliar to the one used by the module, and then use the comparison functions of #47709 to do the comparison. That would probably simplify the comparison logic a lot. But that's also something we can change later on (as long as we change it before 2.8 is released, afterwards we have to watch out for backwards compatibility).

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -253,29 +300,50 @@ def has_different_config(self, net):
-                        different = True
-                        differences.append('ipam_options.%s' % key)
+                for idx, ipam_config in enumerate(self.parameters.ipam_config):
+                    is_ipv4 = re.match(r'^([0-9]{1,3}\.){3}[0-9]{1,3}(/([0-9]|[1-2][0-9]|3[0-2]))?$',
```
You might want to precompile this as a pattern.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -253,29 +300,50 @@ def has_different_config(self, net):
+                            different = True
+                            differences.append('ipam_config[%s].%s' % (idx, key))
+
+        if self.parameters.enable_ipv6 is not None and self.parameters.enable_ipv6 != net['EnableIPv6']:
```
Is `net['EnableIPv6']` always given? Or only for a new enough docker version?

It's probably better to use `net.get('EnableIPv6', False)` instead.

smueller18 replied:
As far as I know, this is always given. But your version sounds better.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -308,6 +377,7 @@ def create_network(self):
                                                   driver=self.parameters.driver,
                                                   options=self.parameters.driver_options,
                                                   ipam=ipam_config,
+                                                  enable_ipv6=self.parameters.enable_ipv6,
```
Could you convert the arguments of `create_network()` into a dict `params`, and pass it using the `**params` syntax? That makes it simpler to only add parameters if they are really needed (like `enable_ipv6`; only pass that if it is not `None`).

smueller18 replied:
Done

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -394,16 +464,18 @@ def main():
         supports_check_mode=True,
-        min_docker_version='1.9.0'
-        # "The docker server >= 1.9.0"
+        min_docker_version='1.10.0'
```
For what precisely is this needed?

smueller18 replied:
see https://github.com/ansible/ansible/pull/47492#discussion_r227458601

felixfontein replied:
If you only add `ipv6_enabled` if it is set to `True`, there's no need to require 1.10.0 then.

smueller18 replied:
I don't know how the behavior is if one defines an IPv6 subnet without setting enable_ipv6 to yes with docker version 1.9.0. There might be errors.

smueller18 replied:
And after checking docker-py 1.9.0 docs, I saw the newly introduced parameter `internal` requires Docker 1.10.0 for externally isolated overlay networks: https://docker-py.readthedocs.io/en/1.9.0/networks.html#network-creation

felixfontein replied:
Ok, then let's keep it. (We can still change that later if we want, before Ansible 2.8 is released. If someone wants to invest the work...)

felixfontein commented on test/integration/targets/docker_network/tasks/tests/ipam.yml:
```diff
@@ -0,0 +1,172 @@
+      - network is changed
+      - network['diff'] | length == 1
+      - network['diff'][0] == "ipam_config[0].subnet"
+
```
You should manually remove the network here if you don't need it later on.

felixfontein commented on test/integration/targets/docker_network/tasks/tests/ipam.yml:
```diff
@@ -0,0 +1,172 @@
+    ipam_config:
+      - subnet: 172.3.27.0/24
+        gateway: 172.3.27.1
+        iprange: 192.168.1.0/24
```
The IP range doesn't lie in the subset. I don't think that's a good idea. It could even be that future versions of docker might fail because of this.

smueller18 replied:
I copied that from the earlier example. Fixing.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -394,16 +464,18 @@ def main():
         ipam_driver=dict(type='str', default=None),
-        ipam_options=dict(type='dict', default={}),
+        ipam_options=dict(type='dict', default={}, removed_in_version='2.12'),
+        ipam_config=dict(type='list', elements='dict', default=[]),
```
Please remove the default `[]`.

smueller18 committed:
docker_network: revert rename of network_four #discussion_r228707101

smueller18 committed:
docker_network: refactor IPAM config comparison #discussion_r228707255, #discussion_r228707280

smueller18 committed:
docker_network: correct spelling of IPv4 and IPv6 #discussion_r228707114, #discussion_r228707138

smueller18 committed:
docker_network: manually remove networks #discussion_r228709051

smueller18 committed:
docker_network: refactor enable_ipv6 condition #discussion_r228707317

smueller18 committed:
docker_network: add mutually_exclusive #discussion_r228707185

smueller18 committed:
docker_network: fix iprange #discussion_r228709072

smueller18 committed:
docker_network: add auxiliary addresses in examples and tests

smueller18 committed:
docker_network: link to docker docs #discussion_r228707018

smueller18 committed:
docker_network: remove list default #discussion_r228707060, #discussion_r228709091

smueller18 committed:
docker_network: introduce params syntax for create_network() #discussion_r228709031

smueller18 commented:
@felixfontein I did all requested changes but https://github.com/ansible/ansible/pull/47492#discussion_r228706928

smueller18 committed:
docker_network: beautify code

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -62,13 +62,33 @@
+    version_added: 2.8
+    description:
+      - List of IPAM config blocks. Consult
+        L(Docker docs, https://docs.docker.com/compose/compose-file/compose-file-v2/#ipam) for valid options and values.
```
Better remove the space after the comma in `L(...,...)`, as you'll otherwise might generate invalid ReST; and having `href=" https://..."` with the space also isn't too good.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -253,29 +316,43 @@ def has_different_config(self, net):
+                            if key == net_key.lower():
+                                camelkey = net_key
+                                break
+                        if not camelkey:
```
Why not change this to
```suggestion
                        if not camelkey or net_config.get(camelkey) != value:
```
and get rid of the second branch?

smueller18 replied:
Sounds good

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -289,26 +366,32 @@ def has_different_config(self, net):
 
     def create_network(self):
         if not self.existing_network:
+            params = dict()
```
You could put `driver` and `options` in here already.

smueller18 replied:
The reason why I did not include those two options is because they have a default. But changing is not a problem. Should I change it to
```
if self.parameters.driver is not None:
    params['driver'] = self.parameters.driver
if self.parameters.driver_options is not None:
   params['options'] = self.parameters.driver_options
```
or without conditionals?
```
params['driver'] = self.parameters.driver
params['options'] = self.parameters.driver_options
```

felixfontein replied:
The second, or simply even do
```
params = dict(
    driver=self.parameters.driver,
    options=self.parameters.driver_options,
)
```

felixfontein commented on lib/ansible/modules/cloud/docker/docker_network.py:
```diff
@@ -394,16 +477,23 @@ def main():
-        ipam_options=dict(type='dict', default={}),
+        ipam_options=dict(type='dict', default={}, removed_in_version='2.12'),
+        ipam_config=dict(type='list', elements='dict', default=None),
+        enable_ipv6=dict(type='bool', default=None),
```
You can also leave `default=None` away. That makes it simpler to parse (IMO). (I don't know why it was already in there for some of the existing options...)

felixfontein reviewed (commented):
Looks good, up to some very minor things :)

felixfontein commented on changelogs/fragments/35370-add_support_for_docker_network_internal_flag.yaml:
```diff
@@ -1,4 +0,0 @@
----
-minor_changes:
-  - "docker_network - ``internal`` is now used to set the ``Internal`` property of the docker network during creation."
```
You should keep that line in this file.

felixfontein commented on changelogs/fragments/docker_network-requirements.yaml:
```diff
@@ -1,6 +1,8 @@
+---
```
This file should keep it's name as well. Only move the version bumps to the new file.

smueller18 replied:
ah, reverting...

felixfontein replied:
Looks good now :) Thanks!

smueller18 commented:
Hope the code is ready for shipment now ;-)

felixfontein commented:
Almost ;) I meant the changelog changes slightly different, see the comments.

felixfontein commented:
(The problem with having one changelog file for all changes is that future PRs will also add themselves to it, which will make it very annoying to backport individual PRs. Backporting is only done for bugfixes, but not for new features.)

smueller18 committed:
docker_network: resolve change requests

felixfontein reviewed (approved):
I think this is ready to go! @DBendit can you take another look?

felixfontein commented:
Meh, now I forgot to:

shipit

smueller18 committed:
docker_network: add yaml header

felixfontein reviewed (approved):
(please don't rebase, then we have to review + shipit again)

shipit

DBendit commented:
LGTM

@smueller18 Thanks for all the work making these changes (and thanks @felixfontein for the thorough reviews) - the final result here looks really clean.

felixfontein commented:
@smueller18 thanks for the PR! It's good to finally have IPv6 support ;)
@DBendit thanks for your review!

About what happens now: first the tests will run through (I hope :) ), this will take some more time since shippable is pretty busy ATM. Afterwards, ansibot should label this PR as `shipit`; and then I'll get someone to merge this.

felixfontein commented:
bot_status

felixfontein commented:
shipit

smueller18 committed:
docker_networking: fix get_ip_version

smueller18 commented:
Damn, while doing my final review I found out that the regex in the function `get_ip_version` allowed IPv4's without netmask... Sorry. Changed it to mandatory and also added a unit test for this function.

felixfontein commented:
No problem :) Now that you started modifying that function, let's extend it a bit more:
1. move the compiled regex to a class-wide constant `CIDR_IPV4` (so it is compiled only once);
2. add another regex `CIDR_IPV6` and test that one as well;
3. if neither of the regex' match, return `not_cidr` (or something else) and fail in that case with an error message (`'"{0}" is neither an IPv4 nor IPv6 CIDR'.format(ipam_config['subnet'])`).

For `CIDR_IPV6` it should be OK to use something simplified, such as `^[0-9a-fA-F:]+/([0-9]|[1-9][0-9]|1[0-2][0-9])$`. If it is invalid, docker will still complain :) The main thing is that it shouldn't match an IPv4 or IPv6 address without subnet.

felixfontein commented:
(And thanks for adding the unit test! That's definitely a good idea!)

smueller18 committed:
docker_network: extend CIDR test

smueller18 commented:
I will check the issue with python 2.6 tomorrow.

felixfontein commented:
That error looks strange, especially as similar constructs appear in other Ansible unit tests. I've asked on #ansible-devel if anyone has seen this error before.

willthames commented on test/units/modules/cloud/docker/test_docker_network.py:
```diff
@@ -0,0 +1,33 @@
+        self.assertEqual(get_ip_version('fdd1:ac8c:0557:7ce2::/64'), 'ipv6')
+        self.assertEqual(get_ip_version('fdd1:ac8c:0557:7ce2::/128'), 'ipv6')
+
+        with self.assertRaises(ValueError) as e:
```
```
self.assertRaises(ValueError, get_ip_version, '192.168.0.1')
```
should work with python 2.6 but you can't check the error message.

smueller18 replied:
In the [test container](quay.io/ansible/default-test-container:1.4.1), the backported [unittest2](https://pypi.org/project/unittest2/) is available. So I added a switch to use the backported unittest2 on python < 2.7

smueller18 committed:
docker_network: use backported unittest2 for python 2.6

smueller18 commented:
All checks have passed ;-)
shipit

felixfontein commented on test/units/modules/cloud/docker/test_docker_network.py:
```diff
@@ -0,0 +1,38 @@
+if sys.version_info < (2, 7):
+    import unittest2 as unittest
+else:
+    import unittest
```
I have just been told that new unit tests must use pytest and not unittest. (That should also solve the Python 2.6 problems...)

smueller18 replied:
I have never used pytest before, but that looks way more elegant

felixfontein replied:
Indeed! *And* it works with Python 2.6 ;)

smueller18 committed:
docker_network: migrate unittest to pytest

felixfontein reviewed (approved):
shipit

DBendit commented on test/integration/targets/docker_network/tasks/tests/ipam.yml:
```diff
@@ -0,0 +1,212 @@
+      - network['diff'] | length == 1
+      - network['diff'][0] == "ipam_config[0].subnet"
+
+- name: Change subnet of network with IPv6 IPAM config
```
[minor] This test's name should call out that it's testing the removal of the netmask. It'll make it easier to figure out what's failing if this test ever finds a regression.

felixfontein replied:
It's not testing removal of the netmask, but produces a failure. The netmask must always be there for it to work. But yes, a more expressive name would be nice. But that can wait until another PR, let's get this one merged first :)

DBendit replied:
It's testing that removing the netmask produces the expected failure, hence why it should have a more expressive name. But as you say, it should be merged, which is why I marked my shipit in another comment on the main thread.

felixfontein replied:
Ah, sorry, I think I misunderstood you :) Anyway, I think we fully agree ;)

DBendit commented:
One very minor thing, but it still looks good enough to throw a shipit on there.

willthames commented:
Merged, thanks @smueller18 for the fix and @DBendit and @felixfontein for the reviews

felixfontein commented:
@smueller18 thanks a lot for this PR!
@DBendit thanks for reviewing this!
@willthames and thanks for merging this!
